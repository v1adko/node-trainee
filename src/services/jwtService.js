import jwt from 'jsonwebtoken';
import configJwt from '../config/jwt';

const { tokenSecret: secret, tokenExpiresIn: expiresIn } = configJwt;

const options = { expiresIn };

class JwtService {
  generateJwt = (user) => {
    const payload = {
      _id: user._id,
      role: user.role
    };
    return jwt.sign(payload, secret, options);
  };

  decoder = (token) => {
    const promise = new Promise((resolve, reject) => {
      jwt.verify(token, secret, (error, decoded) => {
        if (error) {
          reject(new Error('Failed to authenticate token.'));
        } else {
          resolve(decoded);
        }
      });
    });
    return promise;
  };
}

export default new JwtService();
