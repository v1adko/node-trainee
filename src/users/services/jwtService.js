const jwt = require('jsonwebtoken');

const { secretTokkenWord: secret } = require('../config/jwt');

const permissionsConst = require('../config/permissions');

class JwtService {
  constructor() {
    this.generateJwt = (user) => {
      const payload = {
        _id: user._id,
        permissions: Object.values(permissionsConst)
      };
      const options = {
        expiresIn: '2h'
      };

      return jwt.sign(payload, secret, options);
    };
  }
}

module.exports = new JwtService();
