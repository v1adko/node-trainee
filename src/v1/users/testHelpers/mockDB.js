import userDao from '../dao';
import userMocks from './userMocks';
import permissions from '../../../constants/permissions';

const { getUsersFields, user: userFields, admin: adminFields } = userMocks;

class MockDB {
  constructor(DAO) {
    this.DAO = DAO;
  }

  createDefaultUsers = async (userCount = 3) => {
    const users = [];
    getUsersFields(userCount).forEach(async (element) => {
      const { username, password, role } = element;
      const user = await this.createUser(username, password, role);
      users.push(user);
    });
  };

  createDefaultUser = async () => {
    const { username, password, role } = userFields;
    const user = await this.createUser(username, password, role);
    return user;
  };

  createDefaultAdmin = async () => {
    const { username, password, role } = adminFields;
    const user = await this.createUser(username, password, role);
    return user;
  };

  createUser = async (username, password, role = permissions.USER.value) => {
    const createdUser = this.DAO.create(username, password, role);
    return createdUser;
  };

  getAll = async () => {
    const users = await this.DAO.getAll();
    return users;
  };

  cleanDB = async () => {
    await this.DAO.Model.remove();
  };
}

export default new MockDB(userDao);
