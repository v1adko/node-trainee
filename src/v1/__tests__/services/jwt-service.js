import mockDB from '../../../utils/tests-utils/mock-db';
import UserFields from '../../../utils/tests-utils/test-user-fields';
import jwtService from '../../services/jwt-service';

mockDB.connect();

const { username, password, invalidToken } = UserFields;

afterAll(mockDB.closeConnection);

describe('Test the jwt siervice', () => {
  afterEach(mockDB.cleanDB);

  it('should generate token and success decode it', async () => {
    const user = await mockDB.createUser(username, password);
    const token = jwtService.generateJwt(user);

    const tokenData = await jwtService.decoder(token);

    expect(tokenData.id).toBe(user.id);
    expect(tokenData.role).toBe(user.role);
  });

  it('should throw error. whan token is invalid', async () => {
    expect(jwtService.decoder(invalidToken)).rejects.toMatchSnapshot();
  });
});
