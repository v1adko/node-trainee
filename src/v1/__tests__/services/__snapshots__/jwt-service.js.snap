// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test the jwt siervice should throw error. whan token is invalid 1`] = `[TokenValidationError: Invalid token, please repeat authentication.]`;
